xdd: Implement option to load object dictionary from XDD file.

zaporozhets opened the pull request:
Hi All,

This is a WIP merge request that adds support for XDD import to CANopen. I'm still in the process of testing and making improvements, but I’d appreciate any feedback you might have in the meantime.

Feel free to share any questions or suggestions.

Best regards,
Taras

zaporozhets committed:
xdd: Implement option to load object dictionary from XDD file.

Signed-off-by: Taras Zaporozhets <zaporozhets.taras@gmail.com>

acolomb commented:
Thank you very much for this contribution.  Happy to see some progress on the XDD front.  I started taking a look and overall I think it's going in a good direction.

Need to take some more time to put down my thoughts in detail, but some high-level pointers first:

* For new code, we expect to have full type annotations.  The mypy linter run in the PR checks can give you hints about what needs improvement, or best run it locally.  Note that any functions without argument or return type annotations are not even checked by default.
* I don't like the idea of inflating the `canopen.utils` module, even though avoiding code duplication is a worthy goal.  I'll try to come up with a simplification soon.
* Where I do want less duplication is the object code constants declared in both eds.py and xdd.py.  Have a look at commits 7c6dee5a1b040a3de8ea18263f7d8dad9120e076 and db9110e9cd0545b2aa7426243f169d0f863a7640 in master, those can be cherry-picked easily.
* A little more attention to code formatting would be nice.  I use flake8 locally and try to keep it from complaining.  Will report any outstanding issues during more thorough review.
* I assume there is no copyright involved for the `sample.xdd` file?  I don't know where the CANopenNode project got its references from, but just trying to be careful about legal issues with files we include here.

acolomb commented on canopen/objectdictionary/xdd.py:
```diff
@@ -0,0 +1,306 @@
+import re
+import xml.etree.ElementTree as etree
+from configparser import NoOptionError
+from typing import TYPE_CHECKING
```
```suggestion
```

zaporozhets replied:
fixed

acolomb commented on canopen/objectdictionary/xdd.py:
```diff
@@ -0,0 +1,306 @@
+    device_identity = root.find('.//{*}DeviceIdentity')
+    if device_identity is not None:
+        for src_prop, dst_prop, f in [
+            ("vendorName", "vendor_name", lambda val: str(val)),
```
```suggestion
            ("vendorName", "vendor_name", str),
```
This can be simplified in many places, no need for a lambda.

zaporozhets replied:
fixed

acolomb commented on canopen/objectdictionary/xdd.py:
```diff
@@ -0,0 +1,306 @@
+    if device_identity is not None:
+        for src_prop, dst_prop, f in [
+            ("vendorName", "vendor_name", lambda val: str(val)),
+            ("vendorID", "vendor_number", lambda val: int(val, 0)),
```
You could create a `autoint = functools.partial(int, base=0)` constructor once in this module, then use it to simplify in these cases.

zaporozhets replied:
fixed

acolomb reviewed (commented):
Just a partial review again, but I hope you get the general idea... More to come when I get another look with a fresher brain.

tomasbarton-stemcell commented:
hi @zaporozhets, thanks for the work on this PR! Are you still planning to continue with it? I have a project where I could really use xdd/xdc parsing right now, so I'd be glad to pick up where you left off and contribute to this effort if you don't have the bandwidth.

zaporozhets commented:
@acolomb, thank you for the review, and @tomasbarton-stemcell for your interest in this feature.

I was extremely busy last year and didn’t have the chance to address all the comments.

I’m currently on sick leave until Monday, so I finally have some time to catch up. Please give me a few days to refresh my memory and go through all the suggestions. I’ll keep you posted. I’ll definitely appreciate a thorough review here, as Python isn’t really my main area of expertise.

BR,
Taras

zaporozhets commented on canopen/objectdictionary/xdd.py:
```diff
@@ -0,0 +1,306 @@
+
+import re
+import xml.etree.ElementTree as etree
+from configparser import NoOptionError
```
yup. thanks, fixed

zaporozhets commented on canopen/objectdictionary/xdd.py:
```diff
@@ -0,0 +1,306 @@
+import logging
+
```
fixed

zaporozhets commented on canopen/objectdictionary/xdd.py:
```diff
@@ -0,0 +1,306 @@
+from canopen.utils import signed_int_from_hex, calc_bit_length
+
+if TYPE_CHECKING:
+    import canopen.network
```
fixed

zaporozhets commented on canopen/objectdictionary/xdd.py:
```diff
@@ -0,0 +1,306 @@
+RECORD = 9
+
+
+def import_xdd(xdd, node_id):
```
fixed

zaporozhets commented on canopen/objectdictionary/xdd.py:
```diff
@@ -0,0 +1,306 @@
+    _add_dummy_objects_to_od(od, root)
+
+
+    return od
```
fixed

zaporozhets commented on canopen/objectdictionary/xdd.py:
```diff
@@ -0,0 +1,306 @@
+
+    return od
+
+def _add_device_information_to_od(od, root):
```
fixed

zaporozhets commented on canopen/objectdictionary/xdd.py:
```diff
@@ -0,0 +1,306 @@
+                try:
+                    od.node_id = int(node_id, 0)
+                except (ValueError, TypeError):
+                    pass
```
refactored

zaporozhets commented:
Thank you for your patience. I’ve addressed (hopefully all) the comments and refactored several sections to simplify the parsing logic.

1.Added full type annotations and verified the code with `mypy --config-file pyproject.toml ....`
2. Removed changes from `canopen.utils`.
3. Applied updates from `esd.py` to reduce duplicated code.
4. Fixed formatting and validated it with `flake8`.
5. Created a new `sample.xdd` based on `sample.eds`, without relying on third-party EDS/XDD files.
6. Squashed changes in one commit.

@acolomb, @tomasbarton-stemcell could you please review?

tomasbarton-stemcell commented on canopen/objectdictionary/xdd.py:
```diff
@@ -0,0 +1,413 @@
+    else:
+        od.node_id = node_id
+
+    _add_device_information(od, root)
+    _add_object_list(od, root)
+    _add_dummy_objects(od, root)
```
if `xdd` is `None` (which it can be according to the type signature) then this will throw here because root won't be defined.

thisone `root.find('.//{*}DeviceCommissioning')` as well

tomasbarton-stemcell commented on canopen/objectdictionary/xdd.py:
```diff
@@ -0,0 +1,413 @@
+        'USINT': datatypes.UNSIGNED8,
+        'UINT': datatypes.UNSIGNED16,
+        'UDINT': datatypes.UNSIGNED32,
+        'ULINT': datatypes.UNSIGNED32,
```
is this right here? `LINT` above is defined to be 64b, but `ULINT` here is 32b

tomasbarton-stemcell commented on canopen/objectdictionary/xdd.py:
```diff
@@ -0,0 +1,413 @@
+            return int(value, 0)
+
+
+def _convert_variable(
```
unused. Didn't check other files (shouldn't be as this is _), but there is no reference to it in this file.

bizfsc replied:
**`_convert_variable` is defined but never called (dead code).**

It handles `OCTET_STRING`, `DOMAIN`, `VISIBLE_STRING`, `FLOAT_TYPES` etc., but none of the callers use it — `_build_variable` and `_set_parameters_from_xdd_canopen_object` only call `_convert_integer`. This means default values for non-integer types (e.g. `REAL32`, `VISIBLE_STRING`) are silently ignored.

Either wire up `_convert_variable` for non-integer defaults, or remove it.

tomasbarton-stemcell commented on canopen/objectdictionary/xdd.py:
```diff
@@ -0,0 +1,413 @@
+    device_identity = root.find('.//{*}DeviceIdentity')
+    if device_identity is None:
+        raise ValueError("Missing 'DeviceIdentity' section in XDD file")
+    if device_identity is not None:
```
redundant None check (statement above guarantees this is not None).

There is the same problem below: `if general_features is not None:`

bizfsc replied:
**Dead branch: `if device_identity is not None:` is always `True` here.**

The `raise ValueError` on the line above guarantees `device_identity is not None` at this point. The `if` guard is misleading — it implies the block could be skipped. Just dedent the body:

```python
if device_identity is None:
    raise ValueError("Missing 'DeviceIdentity' section in XDD file")
identity_fields = [...]
```

tomasbarton-stemcell commented on canopen/objectdictionary/xdd.py:
```diff
@@ -0,0 +1,413 @@
+            ("granularity", "granularity", autoint, None),
+            ("nrOfRxPDO", "nr_of_RXPDO", autoint, "0"),
+            ("nrOfTxPDO", "nr_of_TXPDO", autoint, "0"),
+            ("bootUpSlave", "simple_boot_up_slave", bool, 0),
```
is `bool` going to work here? I am not sure what the xml parser gives you, but if it is a string "true" or "false" they will both get converted to True using the `bool` function.

I was just going by the example xdd you have in test: `<CANopenGeneralFeatures granularity="8" nrOfRxPDO="4" nrOfTxPDO="4" bootUpSlave="true" />`

bizfsc replied:
**Bug: `bool("false") == True` in Python.**

The XDD attribute value is the string `"true"` or `"false"`. Passing it to `bool(...)` will always return `True` for any non-empty string, including `"false"`.

Fix:

```python
("bootUpSlave", "simple_boot_up_slave", lambda s: str(s).lower() == "true", "false"),
```

tomasbarton-stemcell commented on canopen/objectdictionary/xdd.py:
```diff
@@ -0,0 +1,413 @@
+    value = value.replace(" ", "").upper()
+    if '$NODEID' in value:
+        if node_id is None:
+            logger.warn(
```
`logger.warn` is deprecated: https://docs.python.org/3/library/logging.html#logging.Logger.warning

bizfsc replied:
**`logger.warn()` is deprecated since Python 3.2.**

Use `logger.warning(...)` instead.

tomasbarton-stemcell commented on canopen/objectdictionary/xdd.py:
```diff
@@ -0,0 +1,413 @@
+    var.description = par_tree.get('description', '')
+
+    # Extract data type
+    data_types = {
```
Could probably be a file level constant

tomasbarton-stemcell reviewed (commented):
Hi, thanks for the update! I tested this on a few xdc and xdd files I have in my project (all generated by the editor that comes with the canopennode C library)  and all that I tried worked fine. I added some comments, but I am not really a Python programmer so 🤷🏻‍♂️ 

As a side not, could the `import_od` be relaxed to load xdc files as well? I am not exactly sure what the difference is, but with the files we have I can load them just fine and talk to the devices.

tomasbarton-stemcell commented on canopen/objectdictionary/xdd.py:
```diff
@@ -0,0 +1,413 @@
+        root = etree.parse(xdd).getroot()
+
+    if node_id is None:
+        device_commissioning_elem = root.find('.//{*}DeviceCommissioning')
+        if device_commissioning_elem is not None:
+            node_id_attr = device_commissioning_elem.get("nodeID")
```
One more thing I noticed with my xdc file from CANopenNode editor:

```
<deviceCommissioning NodeID="2" nodeName="scale" actualBaudRate="0" networkNumber="0" networkName="" CANopenManager="false" />
```

The casing is different, it is lower case `deviceCommissioning` and then upper `case NodeID`. I am not sure which is is correct according to the standard.

bizfsc reviewed (commented):
Automated review based on static analysis of the PR diff. Several correctness bugs need fixing before merge; design issues are suggestions.

bizfsc commented on canopen/objectdictionary/xdd.py:
```diff
@@ -0,0 +1,413 @@
+
+logger = logging.getLogger(__name__)
+autoint = functools.partial(int, base=0)
+hex = functools.partial(int, base=16)
```
**Bug: `hex` shadows the `builtins.hex` function.**

`hex` is a Python builtin. Redefining it here makes it unavailable for any downstream code in this module and is confusing.

Suggestion: rename to `parse_hex` or `_from_hex`.

```python
parse_hex = functools.partial(int, base=16)
```

bizfsc commented on canopen/objectdictionary/xdd.py:
```diff
@@ -0,0 +1,413 @@
+) -> ObjectDictionary:
+    od = ObjectDictionary()
+
+    if xdd is not None:
```
**Bug: `NameError` when `xdd=None`.**

If `xdd is None`, `root` is never assigned, but it is used immediately at line 30 (`root.find('.//{*}DeviceCommissioning')`) and again at lines 42-44.

Either raise early or guard all usages:

```python
if xdd is None:
    raise ValueError("source must not be None")
root = etree.parse(xdd).getroot()
```

bizfsc commented on canopen/objectdictionary/xdd.py:
```diff
@@ -0,0 +1,413 @@
+    general_features = root.find('.//{*}CANopenGeneralFeatures')
+    if general_features is None:
+        raise ValueError("Missing 'CANopenGeneralFeatures' element")
+    if general_features is not None:
```
**Same dead-branch issue as with `device_identity`.**

After `raise ValueError("Missing 'CANopenGeneralFeatures' element")`, the `if general_features is not None:` guard is always `True`. Dedent the body.

bizfsc commented on canopen/objectdictionary/xdd.py:
```diff
@@ -0,0 +1,413 @@
+        if object_type == objectcodes.VAR:
+            unique_id_ref = obj.get('uniqueIDRef', None)
+            parameters = root.find(
+                f'.//{{*}}parameter[@uniqueID="{unique_id_ref}"]')
```
**Potential XPath injection / parse error via unsanitised `unique_id_ref`.**

`unique_id_ref` comes directly from an XML attribute. If it contains `"` or `]`, the constructed XPath `'[@uniqueID="..."]'` will either raise a parse error or silently match the wrong element.

The same pattern is repeated for `sub_unique_id` in the ARRAY and RECORD branches.

Mitigation: validate/escape the value, or use a two-step approach (findall + attribute check in Python):

```python
parameters = next(
    (p for p in root.findall('.//{*}parameter')
     if p.get('uniqueID') == unique_id_ref),
    None,
)
```

bizfsc commented on canopen/objectdictionary/xdd.py:
```diff
@@ -0,0 +1,413 @@
+            od.add_object(var)
+
+        # Array
+        elif object_type == objectcodes.ARRAY and sub_number:
```
**Silently skips ARRAY/RECORD objects that have no `subNumber` attribute.**

If `sub_number` is `None` (missing attribute), the whole `elif` is skipped without any warning. The object is silently dropped from the OD. The same applies to the RECORD branch.

Add a `logger.warning(...)` for the unexpected case:

```python
elif object_type == objectcodes.ARRAY:
    if not sub_number:
        logger.warning("ARRAY 0x%04X has no subNumber, skipping", index)
        continue
    ...
```

bizfsc commented on canopen/objectdictionary/xdd.py:
```diff
@@ -0,0 +1,413 @@
+        if entry is None:
+            raise ValueError(
+                "Missing 'entry' attribute for dummy object")
+        p = entry.split('=')
```
**`IndexError` if `entry` contains no `=` character.**

`entry.split('=')` returns a list with one element when `=` is absent; `p[1]` then raises `IndexError`.

Safer:

```python
key, _, raw_value = entry.partition('=')
if not _:
    logger.warning("Unexpected dummyUsage entry format: %r", entry)
    continue
value = int(raw_value, 10)
```

bizfsc commented on canopen/objectdictionary/xdd.py:
```diff
@@ -0,0 +1,413 @@
+    }
+
+    for k, v in data_types.items():
+        if par_tree.find(f'{{*}}{k}') is not None:
```
**Last-wins data-type lookup — missing `break`.**

The loop iterates over all keys even after a match is found. If a `<parameter>` element happened to contain two type tags (malformed XDD), the last one in the dict wins silently.

Add `break` after setting `var.data_type`:

```python
for k, v in data_types.items():
    if par_tree.find(f'{{*}}{k}') is not None:
        var.data_type = v
        break
```

bizfsc commented on test/test_xdd.py:
```diff
@@ -0,0 +1,171 @@
+        ],
+        "int32": [
+            {"hex_str": "7FFFFFFF", "bit_length": 32, "expected": 2147483647},
+            {"hex_str": "80000000", "bit_length": 32, "expected": -2147483648},
```
**Test missing: `bool("false")` bug.**

No test checks that `bootUpSlave="false"` (string) is parsed as `False`. This is the easiest way to catch the `bool(str)` bug in CI.

```python
def test_boot_up_slave_false(self):
    # sample.xdd has bootUpSlave="true"
    self.assertTrue(self.od.device_information.simple_boot_up_slave)
```
And a complementary file with `bootUpSlave="false"` should assert `False`.
